File: InsertionSortList/InsertionSortList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        newhead = ListNode(0)
        newhead.next = head
        pre, cur = newhead, head
        while cur:
        	if cur.next and cur.next.val < cur.val:
        		while pre.next and pre.next.val < cur.next.val:
        			pre = pre.next
        		tmp = pre.next
        		pre.next = cur.next
        		cur.next = cur.next.next
        		pre.next.next = tmp
        		pre = newhead
        	else:
        		cur = cur.next
        return newhead.next

File: InsertionSortList/InsertionSortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* new_list = new ListNode(0);
        ListNode* cur = head;
        while(cur != nullptr) {
        	ListNode* next = cur->next;
        	ListNode* pre = new_list;
        	while(pre->next != nullptr && pre->next->val < cur->val) {
        		pre = pre->next;
        	}
        	cur->next = pre->next;
        	pre->next = cur;
        	cur = next;
        }
        return new_list->next;
    }
};